### DISCUSSION TITLE
Video capture from OneDrive

### DISCUSSION THREAD

**Discussion by AndreyNevolin** [snobu/destreamer#398] [Ideas]
Hi!

Since Microsoft is replacing Stream by OneDrive, do you have plans to implement a tool for grabbing videos from OneDrive?

**Comment by snobu** [comment]
OneDrive? Wasn't it SharePoint?

**Reply by AndreyNevolin** [reply]
I believe, it's both. SharePoint plays a role of a catalog and OneDrive is an actual backend for storing videos. Anyway, whatever it is, the new paradigm is replacing the old one very quickly. All corporate clients of 365 already live under the new paradigm. A tool for capturing video streams from SharePoint/OneDrive would be really handy. I couldn't find any

**Reply by tipicoenlauniFB** [reply]
Now more centers are using SharePoint/OneDrive to share videos but it isn't allowed to download videos, only stream
Header is like https://xxx-my.sharepoint.com/personal/xxxxxxxxxxxxxxx

**Reply by snqqyj** [reply]
Yes. In our company, now the url of the online video for meeting recording is something like below:
https://xxxxxx-my.sharepoint.com/:v:/g/personal/xxxxxx_xxxxxx_xxxxxx_com/EQ5y7SkmWbRDtzS0uE3Rd2sBjxHSkc_GkbzTx9PreCrXig

It seems that the latest SharePoint branch of DeStreamer couldn't support this kind of url yet although @lukaarma recently changed the regex for the url. Hope someone could help update/enhance the SharePoint branch.

**Comment by grzglo** [comment]
Hi all, 
I found the current build (pulled today) working just fine with new links (with SharePoint/Onedrive backend), but it is ignoring "executablePath" value in browser const secition of "/src/TokenCache.ts" file. Sharepoint branch always starts Chromium from "\destreamer\node_modules\puppeteer\.local-chromium\win64-818858\chrome-win". 

Azure AD Conditional Access can be configured to block access from unsupported clients/unmanaged browsers. This way, you can't use Chromium. 

Is there any workaround for this?

**Reply by grzglo** [reply]
I worked around this by replacing "executablePath: getPuppeteerChromiumPath()," with "executablePath: 'C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe'," in files "/src/LoginModules.ts", "/src/TokenCache.ts" and "/src/destreamer.ts". 

@[snobu](https://github.com/snobu) Any chance to make this easier? E.g. Add dependency [edge-paths](https://www.npmjs.com/package/edge-paths). I can update documentation, but first comment if this is right approach to switch from Chromium to Edge.

**Reply by grzglo** [reply]
I was happy to soon, Edge is starting, but download from SharePoint doesn't work. Getting following error message:

[ERROR] Invalid url 'https://{ORGID}.sharepoint.com/personal/{USERID}/_layouts/15/stream.aspx?id={VIDEOID}%2Emp4&ga=1'. 

Can help with tests if anyone have idea how to solve this.

**Reply by mfd** [reply]
Worked solution
1) Run video from SharePoint corporate account or OneDrive, 
2) In Chrome open `Web inspector` and in `Network` filter by `videomanifes`
3) Copy this url

and at terminal run 
`ffmpeg -i "%URL%" -codec copy outputvideo.mp4`

![image](https://user-images.githubusercontent.com/2802926/211963806-67373d1a-413d-405b-825f-1a03be5c3bb6.png)

**Reply by cblf** [reply]
Thank you @mfd , that seems pretty close but this is still not working for me, here is why :
I am copying the "videomanifest" url but this one is very long, especially a very huge "&manifestMetaData" parameter.
The new error is the following :
`[https @ 000001d0bb404a40] overlong headers`

It seems the url contains too much data for ffmpeg or one of its libraries.
Maybe I could remove some parameters in the URL but I don't see what part of it.

Here is a sample of the url parameters (with private information altered) that are producing the problem : 
`videomanifest?provider=spo&inputFormat=mp4&cs=fFNQTw&correlationId=7cd28da0-1043-6000-192b-9e3e5c5dd186&docid=https%3A%2F%2Fmy.sharepoint.com%path%2Fto_my_video%2F_api%2Fv2.0%2Fdrives%2Fb%21t8A85FyVyEO57nbWgmY3z5iHx-OJmoNLq5nxU0lG1duSlOupCWWcSpCdyuAICklZ%2Fitems%2F01BAQH4E6LLDDOKSRJERHIIA2HWJEO2WMI%3Fversion%3DPublished&cTag=%7BE5C658CB-294A-4E24-8403-47B248ED5988%7D%2C12%2C2&action=Access&part=index&format=dash&useScf=True&manifestMetadata=eyJGRm1wZWciOnsiQXVkaW9Db2RlY0luZm8iOiJhYWMiLCJBdWRpb0ZyYW1lU2l6ZSI6MTAyNCwiQXVkaW9TYW1wbGluZ1JhdGUiOjE2MDAwLCJBdWRpb1N0cmVhbUR1cmF0aW9uMTAwTmFubyI6NDE0NTI4MDAwMDAsIkJpdHJhdGUiOjM4MTk0OCwiQ29udGFpbnNBdWRpbyI6dHJ1ZSwiQ29udGFpbnNWaWRlbyI6dHJ1ZSwiQ29udGVudEF3YXJlQml0cmF0ZSI6bnVsbCwiQ29udGVudENhdGVnb3J5IjoidGVhbXNfbWVldGluZyIsIkNyeXB0b2dyYXBoaWNEYXRhQ3JlYXRlZERhdGVVbml4IjoxNjcxNDc1OTc0LCJDcnlwdG9ncmFwaGljRGF0YUlWIjoiMHgyMkZDNUQ4RUQxMUQzOEFFNzYxNkM1OUFFNUJCNTNENyIsIkNyeXB0b2dyYXBoaWNEYXRhS2lkIjoiOTY3YmFmNDU2ZmEuNzE0NDMwLjYxNTc0IiwiRHVyYXRpb24xMDBOYW5vIjo0MTQ1MjgwMDAwMCwiRnJhbWVTaXplIjp7ImhlaWdodCI6MTA4MCwid2lkdGgiOjE5MjB9LCJHb3BTaXplIjo2MDAwMDAwMCwiSDI2NExldmVsIjo0MCwiSGFzQkZyYW1lcyI6ZmFsc2UsIkhlYWRlck9mZnNldEluQnl0ZXMiOjI0LCJIZWFkZXJTaXplIjoxMjk0OTQ4LCJNZXRhZGF0YVNvdXJjZUluZm9ybWF0aW9uIjp7IkF1ZGlvU3RyZWFtRHVyYXRpb25Tb3VyY2UiOjEsIkF1ZGlvU3RyZWFtUmVhZEJ1ZmZlclNpemVJbkJ5dGVzIjoxNjM4NCwiQ29udGFpbmVyRHVyYXRpb25Tb3VyY2UiOjEsIlZpZGVvQml0cmF0ZVNvdXJjZSI6MSwiVmlkZW9TdHJlYW1EdXJhdGlvblNvdXJjZSI6MSwiVmlkZW9TdHJlYW1SZWFkQnVmZmVyU2l6ZUluQnl0ZXMiOjE2Mzg0fSwiUGFja2VkRnJhbWVSYXRlIjp7IkhpZ2giOjU1MjY3NSwiTG93IjozNDU0NH0sIlJlYWxCYXNlRnJhbWVSYXRlIjp7IkhpZ2giOjE2LCJMb3ciOjF9LCJWZXJzaW9uIjp7Il9CdWlsZCI6LTEsIl9NYWpvciI6MywiX01pbm9yIjowLCJfUmV2aXNpb24iOi0xfSwiVmlkZW9Db2RlY0luZm8iOiJoMjY0IiwiVmlkZW9QcmV0cmFuc2NvZGVEYXRhIjpudWxsLCJWaWRlb1N0cmVhbUR1cmF0aW9uMTAwTmFubyI6NDE0NTI4MDAwMDB9LCJNZWRpYUZvdW5kYXRpb24iOnsiQXVkaW9Db2RlY0luZm8iOiJhYWMiLCJBdWRpb0ZyYW1lU2l6ZSI6MCwiQXVkaW9TYW1wbGluZ1JhdGUiOjE2MDAwLCJBdWRpb1N0cmVhbUR1cmF0aW9uMTAwTmFubyI6MCwiQml0cmF0ZSI6MzgxOTQ4LCJDb250YWluc0F1ZGlvIjp0cnVlLCJDb250YWluc1ZpZGVvIjp0cnVlLCJDb250ZW50QXdhcmVCaXRyYXRlIjpudWxsLCJDb250ZW50Q2F0ZWdvcnkiOiJ0ZWFtc19tZWV0aW5nIiwiQ3J5cHRvZ3JhcGhpY0RhdGFDcmVhdGVkRGF0ZVVuaXgiOjE2NzE0NzU5NzQsIkNyeXB0b2dyYXBoaWNEYXRhSVYiOiIweDIyRkM1RDhFRDExRDM4QUU3NjE2QzU5QUU1QkI1M0Q3IiwiQ3J5cHRvZ3JhcGhpY0RhdGFLaWQiOiI5NjdiYWY0NTZmYS43MTQ0MzAuNjE1NzQiLCJEdXJhdGlvbjEwME5hbm8iOjQxNDUyODAwMDAwLCJGcmFtZVNpemUiOnsiaGVpZ2h0IjoxMDgwLCJ3aWR0aCI6MTkyMH0sIkdvcFNpemUiOm51bGwsIkgyNjRMZXZlbCI6NDAsIkhhc0JGcmFtZXMiOm51bGwsIkhlYWRlck9mZnNldEluQnl0ZXMiOjI0LCJIZWFkZXJTaXplIjoxMjk0OTQ4LCJNZXRhZGF0YVNvdXJjZUluZm9ybWF0aW9uIjp7fSwiUGFja2VkRnJhbWVSYXRlIjp7IkhpZ2giOjEyNTAwMDAsIkxvdyI6NzgxMjl9LCJSZWFsQmFzZUZyYW1lUmF0ZSI6bnVsbCwiVmVyc2lvbiI6eyJfQnVpbGQiOi0xLCJfTWFqb3IiOjIsIl9NaW5vciI6MCwiX1JldmlzaW9uIjotMX0sIlZpZGVvQ29kZWNJbmZvIjoiaDI2NCIsIlZpZGVvUHJldHJhbnNjb2RlRGF0YSI6bnVsbCwiVmlkZW9TdHJlYW1EdXJhdGlvbjEwME5hbm8iOjB9fQ%3D%3D&pretranscode=0&transcodeahead=0&enableCdn=1&enhanceAudio=true&psi=3f8026f9-2fdc-4027-8450-1cc88d585945&ccat=1&correlationid=6227efe4-2e5c-438b-b50d-53dfbf55e4f0&hybridPlayback=true&tempauth=eyJ0eXAiOiJKV1QiLCJhbGciOiJub25lIn0.eyJhdWQiOiIwMDAwMDAwMy0wMDAwLTBmZjEtY2UwMC0wMDAwMDAwMDAwMDAvY25hbS1teS5zaGFyZXBvaW50LmNvbUBiMzIzYmNiNC02ZDU4LTRmMjUtODdiZi02MzY2YzNkNjg5YWYiLCJpc3MiOiIwMDAwMDAwMy0wMDAwLTBmZjEtY2UwMC0wMDAwMDAwMDAwMDAiLCJuYmYiOiIxNjc0MTQ3NzIwIiwiZXhwIjoiMTY3NDE2OTMyMCIsImVuZHBvaW50dXJsIjoiTHppQzhON3hpMm9qaTBZUUkxYmtmUFk5bTlWYlR0WjlhYjR3MFBHUmlCTT0iLCJlbmRwb2ludHVybExlbmd0aCI6IjE5MiIsImlzbG9vcGJhY2siOiJUcnVlIiwiY2lkIjoiTjJOa01qaGtZVEF0TVRBME15MDJNREF3TFRFNU1tSXRPV1V6WlRWak5XUmtNVGcyIiwidmVyIjoiaGFzaGVkcHJvb2Z0b2tlbiIsInNpdGVpZCI6IlpUUXpZMk13WWpjdE9UVTFZeTAwTTJNNExXSTVaV1V0Tnpaa05qZ3lOall6TjJObSIsInNpZ25pbl9zdGF0ZSI6IltcImttc2lcIl0iLCJuYW1laWQiOiIwIy5mfG1lbWJlcnNoaXB8YmVub2l0LnRoaWVycnkuYXVkaXRldXJAbGVjbmFtLm5ldCIsIm5paSI6Im1pY3Jvc29mdC5zaGFyZXBvaW50IiwiaXN1c2VyIjoidHJ1ZSIsImNhY2hla2V5IjoiMGguZnxtZW1iZXJzaGlwfDEwMDMzZmZmOTEwZWNmNmFAbGl2ZS5jb20iLCJzaWQiOiI1OTY5ZDIxZi02ZDhkLTRkN2QtODEzOS1kZTJiZWE2YWU0NjMiLCJ0dCI6IjAiLCJ1c2VQZXJzaXN0ZW50Q29va2llIjoiMyIsImlwYWRkciI6IjIwMDE6ODYxOjNjYzI6MWUwMDpjODY5OmEyMjU6Njg5ZTo5M2QifQ.b1c1ZHR6eVBnUDNlWTVQMHU0cmg5MG9vdDZLRHhzU0pLeG53cXdORFJ1dz0`

Thank you for any suggestion.

`

**Reply by cblf** [reply]
Hello again. I finally chose to completely remove the  "&manifestMetaData" parameter, launched the command and it worked plentifully ! :heart:
+1 @mfd , thank you very much !

**Reply by Kungergely** [reply]
@mfd Yes, I can confirm that it works indeed even with fresh recordings too. Thanks a lot for this tip!

**Reply by shebenu** [reply]
looks like videomanifest?provider isn't working anymore

**Reply by mdeepak2021** [reply]
this method no longer works now, videomanifest -returns no result

**Reply by Kungergely** [reply]
Yep, as it turns out the only method that reliably works nowadays is to extract the `videomanifest` URL, trim the part beyond `&format=dash` then plug it into `yt-dlp`, so something in the likes of the following:
```
yt-dlp "https://whatever.com/transform/videomanifest(...)&format=dash" -o "Designated video name.mp4"
```
Please note that `yt-dlp` needs `ffmpeg` in order to be able to merge the (usually separate) audio and video streams into a single file. It goes without saying that `destreamer` isn't working anymore at all.
